aten/src/ATen/native/DilatedConvolution.cpp

```diff
@@ -0,0 +1,879 @@
+    TORCH_CHECK(
+        grad_output.is_contiguous(), "grad_output ought to be contiguous");
+    TORCH_CHECK_DIM_SIZE(grad_output, ndim, dimf, weight.size(0));
+  }
+
+  if (grad_input.defined() || grad_weight.defined() || grad_bias.defined()) {
+    TORCH_CHECK(
+        grad_output.defined(), "grad_output must be defined for gradients");
+  }
+
+  if (columns.defined()) {
+    TORCH_CHECK(columns.is_contiguous(), "columns needs to be contiguous");
+  }
+
+} // conv_dilated_shape_check
+
+/*
+  conv_dilated2d_all_cpu_template
+
+  Main worker. Computes tensors output, grad_input, grad_weight,
+  and/or grad_bias if non-empty.
+ */
+void conv_dilated2d_all_cpu_template(
+    Tensor& output,
+    const Tensor& input_,
+    const Tensor& weight,
+    const Tensor& bias,
+    const Tensor& grad_output_,
+    Tensor& grad_input,
+    Tensor& grad_weight,
+    Tensor& grad_bias,
+    IntArrayRef kernel_size,
+    IntArrayRef stride_size,
+    IntArrayRef pad_size,
+    IntArrayRef dilation_size) {
+  Tensor input = input_;
+  Tensor grad_output = grad_output_;
+
+  // Preliminary shape checks
+  TORCH_CHECK(input.defined(), "input must be defined");
+  TORCH_CHECK(weight.defined(), "weight must be defined");
+  TORCH_CHECK(
+      weight.dim() == 4,
+      "weight must be 4D tensor but got one with shape ",
+      weight.sizes());
+  auto options = input.options();
+
+  // Temporary buffers:
+  Tensor columns = at::empty({0}, options);
+
+  bool is_batch = input.dim() == 3;
+  if (is_batch) {
+    INSERT_BATCH_DIMENSION(input, 1);
+    INSERT_BATCH_DIMENSION(output, 1);
+    INSERT_BATCH_DIMENSION(grad_input, 1);
+    INSERT_BATCH_DIMENSION(grad_output, 1);
+  }
+
+  auto kH = kernel_size[0];
+  auto kW = kernel_size[1];
+  auto dH = stride_size[0];
+  auto dW = stride_size[1];
+  auto padH = pad_size[0];
+  auto padW = pad_size[1];
+  auto dilationH = dilation_size[0];
+  auto dilationW = dilation_size[1];
+
+  int64_t batchSize = input.size(0);
+  int64_t nInputPlane = weight.size(1);
+  int64_t nOutputPlane = weight.size(0);
+  int64_t inputHeight = input.size(2);
+  int64_t inputWidth = input.size(3);
+  int64_t outputHeight = OUTPUTSIZE(0, 2);
+  int64_t outputWidth = OUTPUTSIZE(1, 2);
+
+  // Resize temporary columns
+  if (output.defined() || grad_weight.defined() || grad_input.defined()) {
+    columns.resize_({nInputPlane * kW * kH, outputHeight * outputWidth});
+  }
+  // Initialize
+  if (grad_weight.defined()) {
+    grad_weight.zero_();
+  }
+  if (grad_bias.defined()) {
+    grad_bias.zero_();
+  }
+  if (output.defined() && !bias.defined()) {
+    output.zero_();
+  }
+  // checking shapes after all shapes have been settled:
+  conv_dilated_shape_check(
+      2,
+      input,
+      weight,
+      bias,
+      grad_output,
+      grad_input,
+      grad_weight,
+      grad_bias,
+      kernel_size,
+      stride_size,
+      pad_size,
+      dilation_size,
+      columns);
+
+  // Helpers
+  Tensor input_n = at::empty({0}, options);
```

Comment:
There's no need to allocate an empty tensor (as was the case in the original kernel), because you are using out-of-place `select`. So you are going to allocate a new `input_n` every time you select, and you'll immediately free the zero-size tensor.

Reply:
Fixed.